poll/poll_func.go

```diff
@@ -14,63 +15,85 @@ func PollCmd(w http.ResponseWriter, r *http.Request) {
 	defer r.Body.Close()
 	err := r.ParseForm()
 	if err != nil {
-		log.Print(err)
+		log.Print("Error: ", err)
 		return
 	}
 	poll, err := NewPollRequest(r.Form)
-	if err != nil {
-		log.Print(err)
-		return
-	}
-	if len(Conf.Token) != 0 && Conf.Token != poll.Token {
-		log.Print("Token missmatch. Check you config.json")
-		return
+	var response_type string
+	var response_text string
+	if err == nil {
+		response_type = "in_channel"
+		response_text = poll.Message + ` #poll`
+	} else {
+		response_type = "ephemeral"
+		response_text = err.Error()
 	}
 
-	c := model.NewClient(Conf.Host)
-	c.TeamId = poll.TeamId
+	var response = `{
+		"response_type": "` + response_type + `",
+		"text": "` + response_text + `",
+		"username": "Matterpoll",
+		"icon_url": "https://www.mattermost.org/wp-content/uploads/2016/04/icon.png"
+	}`
+	io.WriteString(w, response)
+	if err == nil {
+		if len(Conf.Token) != 0 && Conf.Token != poll.Token {
+			log.Print("Token missmatch. Check you config.json")
+			return
+		}
 
-	_, err = login(c)
-	if err != nil {
-		log.Print(err)
-		return
-	}
-	p, err := post(c, poll)
-	if err != nil {
-		log.Print(err)
-		return
+		c := model.NewAPIv4Client(Conf.Host)
+		var user *model.User
+		user, err = login(c)
+		if err != nil {
+			log.Print(err)
+			return
+		}
+		go addReaction(c, user, poll)
 	}
-	reaction(c, p, poll)
-	fmt.Fprintf(w, "{'text': 'hello'}")
 }
 
-func login(c *model.Client) (*model.User, error) {
-	r, err := c.Login(Conf.User.Id, Conf.User.Password)
-	if err != nil {
-		return nil, err
+func login(c *model.Client4) (*model.User, error) {
+	u, api_response := c.Login(Conf.User.Id, Conf.User.Password)
+	if api_response != nil && api_response.StatusCode != 200 {
+		return nil, fmt.Errorf("Error: Login failed. API statuscode: ", api_response.StatusCode)
 	}
-	return r.Data.(*model.User), nil
+	return u, nil
 }
 
-func post(c *model.Client, poll *PollRequest) (*model.Post, error) {
-	p := model.Post{
-		ChannelId: poll.ChannelId,
-		Message:   poll.Message + " #poll",
-	}
-	r, err := c.CreatePost(&p)
-	if err != nil {
-		return nil, err
+func addReaction(c *model.Client4, user *model.User, poll *PollRequest) {
+	for try := 0; try < 5; try++ {
```

Comment:
###  commands are executed at the same
We compare the message and confirm its the same. This should be quite safe. We can also check the username of each message, if you think the current approach is not save enough.
### network delay occurs
Usually Matterpoll is run on the same machine as Mattermost. So network problems shouldn't occur.